### DISCUSSION TITLE
vitest has too much memory usage in the face of a large number of benches

### DISCUSSION THREAD

**Discussion by D-Sketon** [vitest-dev/vitest#6535] [Q&A]
I'm a contributor of `es-toolkit`. When I try to run about 200+ `bench.ts` using `yarn vitest bench`, the memory footprint is too high. I allocated 26000mb of memory for the node and it still crashes.
![{5AD4FCF8-ACCA-425C-9A6E-567FEB03F75D}](https://github.com/user-attachments/assets/43507e39-3e1d-4133-9c23-deef8d1c6ca2)
Is there a potential memory leak in vitest, or is there something wrong with the benchmarks I wrote?
benchmark: https://github.com/toss/es-toolkit/tree/main/benchmarks/performance

**Comment by hi-ogawa** [comment]
I tested your `yarn bench` and indeed the memory usage is too much.
Probably this is because we keep all the `samples` array in memory and in your case, this adds up a lot since each bench is very fast and taking many samples (see "samples" column in your table).

There are a few aspects to this issue, For one, Vitest uses `tinybench` underneath and it currently needs accumulating all samples to compute statistics https://github.com/tinylibs/tinybench/issues/61, so keeping samples are necessary at least each bench function level. You can configure `{ time, iterations }` options https://vitest.dev/api/#bench to adjust sample size, but it looks like this is not available as a global config, but maybe we could expose a config like that.

For second, keeping samples during each bench function is not so much of a problem, but a bigger issue is that Vitest is holding all the samples of all benchmarks until the end for reporting purposes.

Not sure what would be the solution, but we probably need to do something about this. Please go ahead creating an issue.

**Comment by AndreyWV** [comment]
Disable `fileParallelism` helped for me and speed up tests extremely
750 files ran in 70s before - and 30s after param change.
Tests with coverage ran 10minutes in 80 files and 1m all files after param change.
Instruction https://vitest.dev/guide/improving-performance
